### ISSUE TITLE
Tests around server-utils

### ISSUE THREAD

**Issue by jrhender** [StampyAI/stampy-ui#191]
Having tests around the `server-utils` directory would be good starting place to add tests as it would help avoid regressions when refactoring and adding new features and provide a framework for testing new features.
The `server-utils` directory is a good place to start because it contains logic in locations like:
- transformation of the markdown to html (in `convertToQuestion`) https://github.com/StampyAI/stampy-ui/blob/d6c6a7b564fd4e81f38234ff0d2322dda762f4e3/app/server-utils/stampy.ts#L215
- kv-cache

**Comment by jrhender** [comment]
@Aprillion @mruwnik , my current vision for this is to cover the "load" in `stampy.ts` functions like `stampy::loadQuestionDetail` and `stampy::loadInitialQuestions`. Does that make sense in your view?

To do this, I think separating the functional code from the requests to coda would be helpful. This would allow the requests to coda to be mocked or injected so that there wouldn't be a need to maintain a coda instance in order to test the functional code.
Mocking could be done with the help of library like Jest Mocks: https://jestjs.io/docs/mock-functions

Any thoughts on the above?

**Comment by Aprillion** [comment]
if you want to start testing functions that talk to the API, I would recommend using https://mswjs.io/ to mock the server responses from Coda, not just mock functions..

**Comment by mruwnik** [comment]
Yup, the fewer and futher back the mocks, the better. Maybe start with setting up the structure for unit tests (plus one or two to make sure it works) and then proceed to testing the coda stuff?

**Comment by jrhender** [comment]
Thanks for the direction @Aprillion and @mruwnik 👍 . Draft PR here: https://github.com/StampyAI/stampy-ui/pull/193
I decided to focus on unit tests for `urlToIframe` as not Coda mocking is required.

**Comment by jrhender** [comment]
With #193 merged, I'll take a look at https://mswjs.io/ for mocking the server responses.

**Comment by jrhender** [comment]
Hi @mruwnik and @Aprillion . I've realized that to test the code in `server-utils`, it would be helpful to mock both Cloudflare KV and Coda responses. My thinking at the moment is to use:
- [msw](https://mswjs.io/) for mocking Coda 
- [miniflare](https://miniflare.dev/get-started/api) for mocking Cloudflare KV

msw was discussed previously but what is your opinion on miniflare?
Though it means an additional dependency (in addition to msw), it seems to me like it could be a good fit to more easily mock the KV calls (e.g. `STAMPY_KV.get(key)`). Otherwise, I think one might need to dig into the internals of the KV client to know which HTTP requests to mock.

**Comment by Aprillion** [comment]
miniflare or something is already used by remix to run localhost, so however that works it can be reused for tests IMHO

**Comment by jrhender** [comment]
Oh yea, you're right, `miniflare` is already a dev dependancy. I hadn't noticed that. I guess it doesn't hurt to give it a go to reuse it for tests then 👍

**Comment by jrhender** [comment]
I found some discussion about testing remix which I'm going through to see if there's anything relevant to stampy-ui's case:
- https://github.com/remix-run/remix/discussions/2481
- https://github.com/remix-run/remix/issues/372

**Comment by jrhender** [comment]
There is a `@remix-run/testing` package that might be useful here: https://www.npmjs.com/package/@remix-run/testing?activeTab=readme

**Comment by jrhender** [comment]
I think the above discussions on remix repo and `@remix-run/testing` are more for testing react components that have dependency on remix, without needing to run remix. As such, I think they are not a relevant to the goal of "tests around server-utils".

I will proceed with the MSW + miniflare approach for now (see draft PR above for the very start of this).

**Comment by jrhender** [comment]
@Aprillion I'm not too far with the implementation (and I'm travelling until the end of May so probably won't make much progress till after then 😅). But I will definitely check out the new MSW release, thanks. I think that direct support for Fetch API could be useful 👍

**Comment by plexish** [comment]
@Aprillion it's JohnH#0114, I added him to team

**Comment by jrhender** [comment]
I've made a bit of progress on this task.
I discovered the Jest environment provided by Miniflare: https://miniflare.dev/testing/jest
I think it could be a promising approach. I have a WIP draft PR here: https://github.com/jrhender/stampy-ui/pull/2

Thus far, the PR is using `undici`, rather than `msw`, to mock the requests to coda. The `msw` mock wasn't working in during my initial attempt to run a test wile `undici` integration is provided out-of-the-box from the miniflare jest env (https://miniflare.dev/testing/jest#mocking-outbound-fetch-requests) and seems to work well. I'm still open to `msw`, but I figure might as well use `undici` as that is presumably tested by the miniflare team for use with their jest env.
